base/dict.jl

```diff
@@ -629,13 +630,28 @@ function pop!(h::Dict)
 end
 
 function _delete!(h::Dict{K,V}, index) where {K,V}
-    @inbounds h.slots[index] = 0x7f
-    @inbounds _unsetindex!(h.keys, index)
-    @inbounds _unsetindex!(h.vals, index)
+    @inbounds begin
+    sz = length(h.slots)
+    nextind = (index & (sz-1)) + 1
+    # if the next slot is empty we don't need a tombstone
+    # and can remove all tombstones that were required by the element we just deleted
+    if isslotempty(h, nextind)
+        while true
+            h.ndel -= 1
+            h.slots[index] = 0x00
+            index = ((index - 2) & (sz-1)) + 1
+            isslotmissing(h, index) || break
+        end
+    else
+        h.slots[index] = 0x7f
+    end
+    _unsetindex!(h.keys, index)
```

Comment:
`_unsetindex!` should be moved before `if isslotempty(h, nextind)` condition because `index` is modified in the first branch.